File: 1-two-sum/two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> needed = {};  // (number, index)
        vector<int> ans;
        int idx = 0;
        for (auto n: nums){
            if (needed.count(target - n) > 0) {
                ans.push_back(idx);
                ans.push_back(needed[target - n]);
                return ans;
            }else{
                needed[n] = idx;
            }
            ++idx;
        }
        ans.push_back(idx-1);
        ans.push_back(needed[target - nums[idx-1]]);
        return ans;
    }
};

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # map
        from collections import defaultdict       
        nums_dict = defaultdict(list)

        for i, v in enumerate(nums):
            nums_dict[v].append(i)

        for k in nums_dict.keys():
            pair = target - k
            if pair in nums_dict:
                if pair != k:
                    return [nums_dict[k][0], nums_dict[pair][0]]
                else:
                    if len(nums_dict[k]) >= 2:
                        return nums_dict[k][:2]
        return []            